### PR TITLE
KAFKA-7536: Initialize TopologyTestDriver with non-null topic

### PR DESCRIPTION
**Pull request by guozhangwang**
In TopologyTestDriver constructor set non-null topic; and in unit test intentionally turn on caching to verify this case.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java [hunk 1]
```diff
@@ -325,7 +325,7 @@ public void onRestoreEnd(final TopicPartition topicPartition, final String store
                 new LogContext()
             );
             globalStateTask.initialize();
-            globalProcessorContext.setRecordContext(new ProcessorRecordContext(0L, -1L, -1, null, new RecordHeaders()));
+            globalProcessorContext.setRecordContext(new ProcessorRecordContext(0, -1L, -1, ProcessorContextImpl.NONEXIST_TOPIC, new RecordHeaders()));
```
Comment (mjsax) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java]: Nit: why change `0L` to `0` ? Timestamp is of type `long`. (sam below)

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 2]
```diff
@@ -779,7 +779,7 @@ private void setup() {
         topology.addStateStore(Stores.keyValueStoreBuilder(
             Stores.inMemoryKeyValueStore("aggStore"),
             Serdes.String(),
-            Serdes.Long()),
+            Serdes.Long()).withCachingEnabled(), // intentionally turn on caching to achieve better test coverage
```
Comment (mjsax) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: I would prefer a separate test: `shouldAllowPrePopulatingStatesStoresWithCachingEnabled()`

**Comment by guozhangwang** [comment]
@mjsax addressed comments.

**Review by bbejeck — APPROVED** [review]
Thanks for this @guozhangwang. LGTM

**Comment by bbejeck** [comment]
failures unrelated

restest this please

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java [hunk 3]
```diff
@@ -889,6 +889,23 @@ private void flushStore() {
...
 
+    @Test
+    public void shouldAllowPrePopulatingStatesStoresWithCachingEnabled() {
+        final Topology topology = new Topology();
+        topology.addSource("sourceProcessor", "input-topic");
+        topology.addProcessor("aggregator", new CustomMaxAggregatorSupplier(), "sourceProcessor");
+        topology.addStateStore(Stores.keyValueStoreBuilder(
+            Stores.inMemoryKeyValueStore("aggStore"),
+            Serdes.String(),
+            Serdes.Long()).withCachingEnabled(), // intentionally turn on caching to achieve better test coverage
```
Comment (mjsax) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java:900]: nit: we can remove this comment, as the test name explains it :)

**Review by mjsax — APPROVED** [review]
One nit.

LGTM. Feel free to merge.

Test timed out. Retest this please.

**Comment by guozhangwang** [comment]
I tried to cherry-pick to the earliest branch (1.1) but realized it has a few other diffs like KAFKA-6967 in the middle which are not cherry-picked either and hence making the divergence too large. Instead I'd only cherry-pick to 2.0 and 2.1.

**Pull request merged**